google/cloud/internal/timer_queue.cc

```diff
@@ -46,80 +56,83 @@ void TimerQueue::Shutdown() {
+    // wake up and recompute the sleep time. But note that the leader thread
+    // does not need to relinquish that role to do so.
+    auto until = timers_.begin()->first;
+    auto predicate = [this, until] {
+      return shutdown_ || timers_.empty() || timers_.begin()->first != until;
     };
-    // TODO(#10512): Only wake one thread to service the next timer instead
-    // of all of them.
-    cv_.wait_until(lk, until, predicate);
-    if (shutdown_) return NotifyShutdown(std::move(lk));
-    auto now = std::chrono::system_clock::now();
-    if (HasExpiredTimer(lk, now)) {
-      ExpireTimer(std::move(lk), now);
+    if (cv_.wait_until(lk, until, predicate)) {
+      // Timers can be expired only if the CV exits due to a timeout.
+      continue;
     }
+    // if we get here we know that predicate() is false. Which implies that
+    // timers_ is not empty.
+    auto p = std::move(timers_.begin()->second);
+    timers_.erase(timers_.begin());
+    // Relinquish the leader role, release the mutex and then signal a follower.
+    has_leader_ = false;
+    is_leader = false;
+    lk.unlock();
+    // Elect a new leader (if available) to continue expiring timers.
+    cv_follower_.notify_one();
+    // This may run user code (in the continuations for the future). That may
+    // do all kinds of things, including calling back on this class to create
+    // new timers. We cannot hold the mutex while it is running.
+    p.set_value(std::chrono::system_clock::now());
+    lk.lock();
   }
+  CancelAll(std::move(lk), "TimerQueue shutdown");
 }
 
 void TimerQueue::CancelAll() {
-  std::multimap<std::chrono::system_clock::time_point, PromiseType> timers;
-  {
-    std::lock_guard<std::mutex> lk(mu_);
-    std::swap(timers, timers_);
-  }
-  for (auto& timer : timers) {
-    timer.second.set_value(Status(StatusCode::kCancelled, "Timer cancelled"));
-  }
+  CancelAll(std::unique_lock<std::mutex>(mu_), "TimerQueue cancel all");
 }
 
-// Requires a lock to be held before calling.
-bool TimerQueue::HasExpiredTimer(std::unique_lock<std::mutex> const&,
-                                 std::chrono::system_clock::time_point tp) {
-  return !timers_.empty() && timers_.begin()->first < tp;
-}
-
-void TimerQueue::ExpireTimer(std::unique_lock<std::mutex> lk,
-                             std::chrono::system_clock::time_point tp) {
-  auto p = std::move(timers_.begin()->second);
-  timers_.erase(timers_.begin());
-  auto const has_expired_timers = HasExpiredTimer(lk, tp);
-  lk.unlock();
-  // We do not want to expire all the timers in this thread, as this would
-  // serialize the expiration.  Instead, we pick another thread to continue
-  // expiring timers.
-  if (has_expired_timers) cv_.notify_one();
-  p.set_value(tp);
-}
-
-void TimerQueue::NotifyShutdown(std::unique_lock<std::mutex> lk) {
-  auto timers = std::move(timers_);
+void TimerQueue::CancelAll(std::unique_lock<std::mutex> lk, char const* msg) {
+  decltype(timers_) timers;
```

Comment:
Nit: Declaring this outside the lock was previously a feature, but I guess it doesn't really matter.

Reply:
N/A